docs/proposals/error-handling-guideline.md

```diff
@@ -0,0 +1,279 @@
+
+## Error output recommendation
+
+### Dos
+
+- Provide full description if the user input does not match what ORAS CLI expected. A full description should include the actual input received from the user and expected input
+- Use the capital letter ahead of each line of any error message
+- Print human readable error message. If the error message is mainly from the server and varies by different servers, tell users that the error response is from server. This implies that users may need to contact server side for troubleshooting.
+- Provide specific and actionable prompt message with argument suggestion or show the example usage for reference. (e.g, Instead of showing flag or argument options is missing, please provide available argument options and guide users to "--help" to view more examples)
+- If the actionable prompt message is too long to show in the CLI output, consider guide users to ORAS user guide or troubleshooting guide with the permanent link.
+- If the error message is not enough for troubleshooting, guide users to use "--verbose" to print much more detailed logs
+
+### Don'Ts
+
+- Do not use a formula-like or a programming expression in the error message. (e.g, `json: cannot unmarshal string into Go value of type map[string]map[string]string.`, or `Parameter 'xyz' must conform to the following pattern: '^[-\\w\\._\\(\\)]+$'`)
+- Do not use ambiguous expressions which mean nothing to users. (e.g, `Something unexpected happens`, or `Error: accepts 2 arg(s), received 0`)
+- Do not print irrelevant error message to make the output noisy. The more irrelevant output you produce, the longer it's going to take the user to figure out what they did wrong.
+
+## How to write friendly error message
+
+### Recommended error message structure
+
+Here is a sample structure of an error message:
+
+```text
+Error: {Error|Error response from registry}: {Error description}
+[Usage: {Command usage}]
+[{Recommended solution}]
+```
+
+- Status code is an optional information. Printed out the status code if the error message is generated from the server side. 
+- Command usage is also an optional information but it's recommended to be printed out when user input doesn't follow the standard usage or examples.
+- Recommended solution (if any) should follow the general guiding principles described above.
+
+### Examples
+
+Here are some examples of writing error message with helpful prompt actionable information:
+
+#### Example 1: When no reference provided in `oras copy`
+
+Current behavior and output:
+
+```
+$ oras cp
+Error: accepts 2 arg(s), received 0
+```
+
+Suggested error message:
+
+```
+$ oras cp
+Error: "oras copy" requires exactly 2 arguments but received 0.
+Usage: oras copy [flags] <from>{:<tag>|@<digest>} <to>[:<tag>[,<tag>][...]]
+Please specify 2 arguments as source and destination respectively. Run "oras copy -h" for more options and examples
+```
+
+#### Example 2: When reference does not match with the expected format
```

Comment:
This is not an example of `reference does not match with the expected format`, this is a well-known mistake of using `repo show-tags` as `tag list`
```suggestion
#### Example 2: When user mistakenly use `tag list` command instead of `repo show-tags`
```